### PR TITLE
Update Load replica job

### PR DESCRIPTION
**Pull request by jja725**
### What changes are proposed in this pull request?

Update load job to adopt multi replicas

### Why are the changes needed?
part of PR to support loading multi replicas

### Does this PR introduce any user facing changes?
new load option `replicas`

### CODE REVIEW

File: dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java [hunk 1]
```diff
@@ -61,7 +61,7 @@ public List<BlockWorkerInfo> getPreferredWorkers(List<BlockWorkerInfo> blockWork
     List<BlockWorkerInfo> workers = HASH_PROVIDER.getMultiple(fileId, count);
     if (workers.size() != count) {
       throw new ResourceExhaustedException(String.format(
-          "Found %d workers from the hash ring but %d required", blockWorkerInfos.size(), count));
+          "Found %d workers from the hash ring but %d required", workers.size(), count));
```
Comment (lucyge2022) [inline dora/core/client/fs/src/main/java/alluxio/client/file/dora/ConsistentHashPolicy.java:64]: good catch

File: dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java [hunk 2]
```diff
@@ -13,23 +13,31 @@
...
-    int nextWorkerIdx = Math.floorMod(mCounter.incrementAndGet(), workerInfos.size());
-    WorkerInfo pickedWorker = workerInfos.toArray(new WorkerInfo[workerInfos.size()])
-        [nextWorkerIdx];
-    return pickedWorker;
+    ArrayList workers = new ArrayList<>();
+    for (int i = 0; i < count; i++) {
+      int nextWorkerIdx = Math.floorMod(mCounter.incrementAndGet(), workerInfos.size());
+      WorkerInfo pickedWorker =
+          workerInfos.toArray(new WorkerInfo[workerInfos.size()])[nextWorkerIdx];
+      workers.add(pickedWorker);
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java:39]: do deduplication , i m assuming when we call pickWorkers we always wants to be pick different "count" of workers right? need to call it out in interface doc if thats the case

File: dora/core/server/master/src/main/java/alluxio/master/job/WorkerAssignPolicy.java [hunk 3]
```diff
@@ -11,23 +11,28 @@
...
+public interface WorkerAssignPolicy {
 
   /**
-   * Pick a worker based on the policy.
-   * @param object object
+   * Pick workers based on the policy. If the expected number is not met, it will return empty list
+   *
+   * @param object      object
    * @param workerInfos worker information
+   * @param count       expected number of workers
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/WorkerAssignPolicy.java:33]: expected num of different of workers?

File: dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java (hunk 2 above)
Comment (jja725) [reply dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java:39]: In worker location policy there's dedup so no need to do it here

File: dora/core/server/master/src/main/java/alluxio/master/job/CopyJob.java [hunk 4]
```diff
@@ -302,9 +302,14 @@ public List<CopyTask> getNextTasks(Collection<WorkerInfo> workers) {
       if (routes.isEmpty()) {
         return Collections.unmodifiableList(tasks);
       }
-      WorkerInfo workerInfo = mWorkerAssignPolicy.pickAWorker(StringUtil.EMPTY_STRING, workers);
+      List<WorkerInfo>
+          workerInfo = mWorkerAssignPolicy.pickWorkers(StringUtil.EMPTY_STRING, workers, 1);
+      if (workerInfo.isEmpty()) {
+        LOG.warn("No worker available for copy job {}.", mJobId);
+        mRetryRoutes.addAll(routes);
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/CopyJob.java:309]: bail here? otherwise workerInfo.get(0) has index out of bound

Comment (jja725) [reply dora/core/server/master/src/main/java/alluxio/master/job/CopyJob.java:309]: fixed

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java [hunk 5]
```diff
@@ -274,7 +267,23 @@ private List<LoadSubTask> createSubTasks(UfsStatus ufsStatus) {
     else {
       subTasks.add(new LoadDataSubTask(ufsStatus, mVirtualBlockSize, 0, contentLength));
     }
-    return subTasks;
+    List<LoadSubTask> subTasksWithWorker = assignSubtasksToWorkers(subTasks, workers, mNumReplica);
```
Comment (lucyge2022) [inline dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:270]: so replica also determines num of replica of metadata right?

File: dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java (hunk 2 above)
Comment (lucyge2022) [reply dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java:41]: no this is roundrobin

File: dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java (hunk 5 above)
Comment (jja725) [reply dora/core/server/master/src/main/java/alluxio/master/job/DoraLoadJob.java:270]: yes

File: dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java (hunk 2 above)
Comment (jja725) [reply dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java:41]: oh, so if workerinfos are distinct workers then this round robin should return distinct workers I think? I would update the method signature

Comment (lucyge2022) [reply dora/core/server/master/src/main/java/alluxio/master/job/RoundRobinWorkerAssignPolicy.java:41]: oh i get what u r saying, u r right here.

**Review by lucyge2022 — APPROVED** [review]
LGTM

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Comment by jja725** [comment]
alluxio-bot, merge this please

**Pull request merged**